reactioncommerce/reaction
# Add a way to extend catalog product publication

Resolves #4725 
Impact: **minor**  
Type: **feature**

Related documentation updates: https://github.com/reactioncommerce/reaction-docs/pull/699

## Changes
1. A plugin can now include a `catalog` object in `registerPackage`, with `customPublishedProductFields` and `customPublishedProductVariantFields` that are set to arrays of property names. These will be appended to the core list of fields for which published status should be tracked. This is used to build the hashes that are used to display an indicator when changes need to be published.
1. A plugin can now use the `functionsByType` pattern to register one or more functions of type "publishProductToCatalog", which are called with `(catalogProduct, { context, product, shop, variants })` and expected to mutate `catalogProduct` if necessary.
1. In support of the above, fine-tuned and made more explicit the list of variant fields that need publishing.
1. By reviewing the list of published properties, this fixes some issues where changing certain things about a product or variant, such as media changes, did not cause the "needs publishing" indicator to appear.
1. Optimized the publish-to-catalog code to hit the database fewer times.

## Breaking changes
None

## Testing
1. Verify that changing published product fields as an operator causes the "needs publishing" indicator to appear, and then verify that publishing your change works.
1. Try creating a plugin with the new API and verify that changes to your fields show the "needs publishing" indicator, and then those changes make it into the catalog.
